dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/dataintegrity/DefaultDataIntegrityService.java

```diff
@@ -784,38 +802,95 @@ private String getInvalidExpressionDescription( String expression )
     }
 
     @Override
-    @Transactional( readOnly = true )
-    public Map<String, DataIntegritySummary> getSummaries( Set<String> checks, JobProgress progress )
+    public Map<String, DataIntegritySummary> getSummaries( Set<String> checks, long timeout )
     {
-        return runDataIntegrityChecks( "Data Integrity summary checks", expandChecks( checks ), progress,
-            check -> check.getRunSummaryCheck().apply( check ) );
+        return getCached( checks, timeout, summaryCache );
     }
 
+    // OBS! We intentionally do not open the transaction here to have each check
+    // be independent
     @Override
-    @Transactional( readOnly = true )
-    public Map<String, DataIntegrityDetails> getDetails( Set<String> checks, JobProgress progress )
+    public void runSummaryChecks( Set<String> checks, JobProgress progress )
+    {
+        runDataIntegrityChecks( "Data Integrity summary checks", expandChecks( checks ), progress, summaryCache,
+            check -> check.getRunSummaryCheck().apply( check ),
+            ( check, ex ) -> new DataIntegritySummary( check, new Date(), ex.getMessage(), -1, null ) );
+    }
+
+    @Override
+    public Map<String, DataIntegrityDetails> getDetails( Set<String> checks, long timeout )
     {
-        return runDataIntegrityChecks( "Data Integrity details checks", expandChecks( checks ), progress,
-            check -> check.getRunDetailsCheck().apply( check ) );
+        return getCached( checks, timeout, detailsCache );
     }
 
-    private <T> Map<String, T> runDataIntegrityChecks( String stageDesc, Set<String> checks, JobProgress progress,
-        Function<DataIntegrityCheck, T> runCheck )
+    // OBS! We intentionally do not open the transaction here to have each check
+    // be independent
+    @Override
+    public void runDetailsChecks( Set<String> checks, JobProgress progress )
+    {
+        runDataIntegrityChecks( "Data Integrity details checks", expandChecks( checks ), progress, detailsCache,
+            check -> check.getRunDetailsCheck().apply( check ),
+            ( check, ex ) -> new DataIntegrityDetails( check, new Date(), ex.getMessage(), List.of() ) );
+    }
+
+    private <T> Map<String, T> getCached( Set<String> checks, long timeout, Cache<T> cache )
+    {
+        Set<String> names = expandChecks( checks );
+        long giveUpTime = currentTimeMillis() + timeout;
+        Map<String, T> resByName = new LinkedHashMap<>();
+        boolean retry = false;
+        do
+        {
+            if ( retry )
```

Comment:
This part seems a bit complicated. Just trying to understand the reason/logic for the retry?

Reply:
The loop tries to get values from the cache. I assume you just triggered a set of checks moments ago so you will not immediately find all results in the cache. Therefore you can give the `timeout` value. When we have extracted results once from the cache we check if we have all results we wanted, if not and if we have not reached the timeout point we do sleep for a while and then try again to find the missing results in the cache. Again, if not complete sleep and retry. Until either we found all results in the cache or the timeout stops us. This makes it more convinient for clients to trigger and fetch the data. And gives us the chance to use block waiting in the future where we can wait exactly until all results are present in the cache (or some timeout occurred).